src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/NativeMemorySegmentImpl.java

```diff
@@ -98,21 +98,28 @@ public static MemorySegment makeNativeSegment(long bytesSize, long alignmentByte
             unsafe.setMemory(buf, alignedSize, (byte)0);
         }
         long alignedBuf = Utils.alignUp(buf, alignmentBytes);
-        MemoryScope scope = MemoryScope.createConfined(null, () -> {
+        AbstractMemorySegmentImpl segment = new NativeMemorySegmentImpl(buf, alignedSize,
+                defaultAccessModes(alignedSize), scope);
+        scope.addOrCleanupIfFail(new ResourceList.ResourceCleanup() {
+            @Override
+            public void cleanup() {
                 unsafe.freeMemory(buf);
                 nioAccess.unreserveMemory(alignedSize, bytesSize);
-            }, null);
-        MemorySegment segment = new NativeMemorySegmentImpl(buf, alignedSize,
-                defaultAccessModes(alignedSize), scope);
+            }
+        });
         if (alignedSize != bytesSize) {
             long delta = alignedBuf - buf;
             segment = segment.asSlice(delta, bytesSize);
         }
         return segment;
     }
 
-    public static MemorySegment makeNativeSegmentUnchecked(MemoryAddress min, long bytesSize, Runnable cleanupAction, Object ref) {
-        return new NativeMemorySegmentImpl(min.toRawLongValue(), bytesSize, defaultAccessModes(bytesSize),
-                MemoryScope.createConfined(ref, cleanupAction == null ? MemoryScope.DUMMY_CLEANUP_ACTION : cleanupAction, null));
+    public static MemorySegment makeNativeSegmentUnchecked(MemoryAddress min, long bytesSize, Runnable cleanupAction, MemoryScope scope) {
+        scope.checkValidStateSlow();
+        AbstractMemorySegmentImpl segment = new NativeMemorySegmentImpl(min.toRawLongValue(), bytesSize, defaultAccessModes(bytesSize), scope);
+        if (cleanupAction != null) {
+            scope.addOnClose(cleanupAction);
```

Comment:
This is a bit controversial: should the cleanup action be called if there's a race and, by the time we create the unsafe segment, the scope is already closed? Normally, we call `addOrCleanupIfFail` because we don't want e.g. `MemorySegment::allocateNative` to leave resource around if the scope happens to be closed - but in this case? I think I stick by the current decision - otherwise we're essentially making a guarantee that the cleanup action provided by the user will always be called, which I'm not sure I wanna make.